[docs] runtime version doc page (#13819)
expo/expo: docs/pages/distribution/runtime-versions.md

---
title: Runtime Versions
---

> Custom runtime versions are not supported on the classic build system (`expo build`); these apps will always use the SDK version as the basis for determining runtime compatibility. 

> Over-the-air updates with `expo-updates` will only work in binaries with a [compatible](../workflow/publishing/#what-version-of-the-app-will-my) runtime. By default, the runtime version is the Expo SDK version: the runtime is the set of libraries in the Expo SDK. Apps built with EAS Build will include only the dependencies that are present at the time that the app binary is built, and so the Expo SDK version will often not properly describe the runtime of the app. In these cases, you can start using `runtimeVersion` and it will take precedence over the Expo SDK version.

Reviewer: ```suggestion
Over-the-air updates with `expo-updates` are only applied to binaries with a [compatible](../workflow/publishing/#what-version-of-the-app-will-my) runtime. By default, the runtime version is the Expo SDK version: the runtime is the set of libraries in the Expo SDK. If you choose to customize your runtime by adding a module with native code with [Config Plugins](../guides/config-plugins.md) or removing dependencies that are not referenced in your project with EAS Build, the Expo SDK version will no longer properly describe the runtime of the app. In these cases, you should specify a `runtimeVersion` to ensure your updates are only delivered to compatible clients.
```

Reviewer: @brentvatne I couldn't find documentation for removing modules not referenced in teh project with EAS build.  Would be nice to link to it here if it exists

Reviewer 2: * "work in" is more direct and less abstract than "applied to"
* all this applies to modules without prebuild config plugins. what matters is the js–native interface

```suggestion
Over-the-air updates with `expo-updates` work only in apps with a [compatible](../workflow/publishing/#what-version-of-the-app-will-my) runtime. By default, the runtime version is the Expo SDK version: the runtime is the set of libraries in the Expo SDK. If you choose to customize your runtime by adding or removing modules with native code, the Expo SDK version will no longer properly describe the runtime of the app. In these cases, you need to specify a `runtimeVersion` to ensure your updates are delivered only to compatible clients.  This `runtimeVersion` should be updated whenever you update your project's native modules and change the JS–native interface.
```

Reviewer 3: i didn't add that point here, i'm not sure it makes sense to include the text "or removing dependencies that are not referenced in your project with EAS Build"

Reviewer: To me, "work in" seems hand wavy and imprecise, but I'm not very opinionated

I was trying to preserve the point in the existing text but happy to cut it

> Over-the-air updates with `expo-updates` work only in apps with a [compatible](../workflow/publishing/#what-version-of-the-app-will-my) runtime. By default, the runtime version is the Expo SDK version: the runtime is the set of libraries in the Expo SDK. If you choose to customize your runtime by adding or removing modules with native code, the Expo SDK version will no longer properly describe the runtime of the app. In these cases, you need to specify a `runtimeVersion` to ensure your updates are delivered only to compatible clients.  This `runtimeVersion` should be updated whenever you update your project's native modules and change the JS–native interface.

Reviewer 3: ```suggestion
Every over-the-air update with `expo-updates` targets one [compatible](../workflow/publishing/#what-version-of-the-app-will-my) runtime. Each time you build a binary for your app it includes the native code present at the time of the build and only that code, and this unique combination and configuration of the build is what is represented by the runtime version.

By default, the runtime version is determined by the Expo SDK version, but this will not adequately describe the different runtime versions of your app if you build more than once per SDK release. In this case, you will need to specify a `runtimeVersion` to ensure your updates are delivered only to compatible builds. This `runtimeVersion` should be updated whenever you update your project's native modules and/or change the JS–native interface.
```

Author: I like this change, @ide @tcdavis have any objections?

Reviewer 2: Every -> Each reads better IMO (you could read "every" as meaning all of the updates target the same RTV).

I also think "and" is better than "and/or" here since changing native modules doesn't always require rebuilding manifests.

> A runtime version must conform to this [description](../versions/latest/config/app/#runtimeversion)

Author: Right now, this link does not display the regex we use to validate the runtime version.

I put up a PR to add a description field here: https://github.com/expo/universe/pull/7959

How do I update the local XDL with this update? Is there a script?

Author: Yes: `yarn run schema-sync <SDK version integer>`. Will run once https://github.com/expo/universe/pull/7959 lands.

> A runtime version must conform to this [description](../versions/latest/config/app/#runtimeversion)

Reviewer 3: ```suggestion
The runtime version string must conform to [this format](/versions/latest/config/app.md#runtimeversion).
```

can we inline the human regex description here? it seems worthwhile to avoid having to jump to another page. we could possibly make a react component that allows us to refer to the runtimeVersion in the schema to print it here so that we don't have to just copy and paste it

Author: I like the inlining, making an issue out of it.

## Setting the runtime version for an update

> [Updates](/workflow/publishing.md#how-to-publish) published with the runtime version set in the `app.json` will be delivered to builds running the same runtime version, and only to those builds.

Reviewer 3: ```suggestion
[Updates](/workflow/publishing.md#how-to-publish) published with the runtime version set in `app.json` will be delivered to builds running the same runtime version, and only to those builds.
```

```json
{
  "expo": {
    "runtimeVersion": "2.718"
  }
}
```
## Setting the runtime version for a build

### Configuration for the managed workflow

> If you are using the [managed workflow](../introduction/managed-vs-bare/#managed-workflow), than specifying the runtime version in your `app.json` is sufficient.

Reviewer 3: ```suggestion
If you are using the [managed workflow](../introduction/managed-vs-bare/#managed-workflow), `runtimeVersion` is specified in `app.json`:
```

```json
{
  "expo": {
    "runtimeVersion": "2.718"
  }
}
```

### Configuration for the bare workflow

> If you are using the [bare workflow](../introduction/managed-vs-bare/#bare-workflow), you need to edit `Expo.plist` on iOS and `AndroidManifest.xml` on Android.

Reviewer 3: ```suggestion

If you are using the [bare workflow](/introduction/managed-vs-bare.md#bare-workflow), set the runtime version in `Expo.plist` on iOS and `AndroidManifest.xml` on Android.
```

> For an iOS build, add an entry to the `Expo.plist` whose key is `EXUpdatesRuntimeVersion` and value is a string set to the desired runtime version. 

Reviewer 3: ```suggestion
For an iOS build, add an entry to the `Expo.plist` with the key `EXUpdatesRuntimeVersion`. The value is a string that represents the runtime version. 
```

```diff 
+ <key>EXUpdatesRuntimeVersion</key>
+ <string>2.718</string>
```

> For an Android build, add a `<meta-data>` element to the `AndroidManifest.xml` whose `android:name` attribute is `expo.modules.updates.EXPO_RUNTIME_VERSION` and `android:value` attribute is the desired runtime version.

Reviewer 3: ```suggestion
For an Android build, add a `<meta-data>` element to the `AndroidManifest.xml` whose `android:name` attribute is `expo.modules.updates.EXPO_RUNTIME_VERSION` and `android:value` attribute is a string that represents the desired runtime version.
```

```json
{
  "expo": {
    "runtimeVersion": "2.718"
  }
}
```

### Configuration for the bare workflow

If you are using the [bare workflow](/introduction/managed-vs-bare.md#bare-workflow), set the runtime version in `Expo.plist` on iOS and `AndroidManifest.xml` on Android.

For an iOS build, add an entry to the `Expo.plist` with the key `EXUpdatesRuntimeVersion`. The value is a string that represents the runtime version. 

```diff 
+ <key>EXUpdatesRuntimeVersion</key>
+ <string>2.718</string>
```
For an Android build, add a `<meta-data>` element to the `AndroidManifest.xml` whose `android:name` attribute is `expo.modules.updates.EXPO_RUNTIME_VERSION` and `android:value` attribute is a string that represents the desired runtime version.

```diff
+ <meta-data android:name="expo.modules.updates.EXPO_RUNTIME_VERSION" android:value="2.718"/>
```

## FAQ

### Can I have a different runtime version on iOS and Android?

Yes, if you want to be able to control the runtime version on a platform level, you can:
1. Have platform specific release channels: `ios-production`, `android-production`.

> 2. Have platform specific runtime versions: `ios-1.0.0, `android-1.0.0.

Reviewer 3: you still need to set this on the root `runtimeVersion` field in app.json right?

Reviewer 2: ```suggestion
2. Have platform specific runtime versions: `ios-1.0.0, `android-1.0.0`.
```
